docs/decisions/0005-access-policy-locks.rst

```diff
@@ -0,0 +1,72 @@
+5. Access Policy Redemption Locks
+#################################
+
+Status
+******
+
+**Proposed**
+
+March 2023
+
+
+Context
+*******
+We've introduced a new Django service, ``enterprise-subsidy``, to provide a new implementation of Learner Credit,
+in which learners in an enterprise can redeem the balance of their enterprise's Learner Credit ledger to pay
+for verified enrollments in any kind of content supported via Enterprise Catalogs.
+See the `Access Policy Functionality ADR`_ for more context.
+
+We need to treat redemptions of learner credit value via an access policy as a `shared resource`,
+since they support limits on how much value can be spent, either "per-learner" (i.e. no learner covered by
+a policy may spend more than some dollar or enrollment limit), or "per-policy" (i.e. no more than
+some dollar or enrollment limit may be consumed in aggregate, across all learners, via the policy).
+
+Decision
+********
+We'll use a distributed mechanism to lock policies during a redemption request.  See
+the `Ledger Distributed Locking ADR`_ for more details. Note that we'll initially
+rely on memcached (via Django's cache API) instead of Redis as our distributed store of locks.
+
+Consider the following types of access policies:
+
+- Per-policy spend/enrollment limited: in these policies, value used through
+  the policy can only succeed up to some limit; the limit applies `in aggregate` across all transactions
+  originating from the policy.  Here, we want to lock the entire policy during redemption;
+  no concurrent redemption should be allowed, because the limit on value redemption must take into
+  account `all` transactions which originate from the policy.
+- Per-learner spend/enrollment limited: in these policies, redemptions for `any user` in the policy
+  may only succeed up to some limit.  Here, we want to lock `at least` at the policy level, which means
+  no concurrent redemption is allowed by any other transaction (or for any other user) should be allowed.
+  Ideally, we'd lock on both the policy ``uuid`` and the learner identifier, so that we are not too
+  restrictive.
+- Combination per-policy and per-learner limited: This policy type is the union of the first two:
+  no more than some policy-level limit can be spent in aggregate, and no individual learner in the policy
+  may spend more than some learner-level limit.  In this case, we `must` lock the entire policy during
+  redemption, because again, the limit on value redemption must take into account `all` transactions
+  which originate from the policy.
+
+During redemption, we'll acquire a lock key that is scoped to the ``uuid`` of the policy through
+which redemption occurs.  This means that no other, concurrent request can redeem through this policy
+while this lock is held.  This is the simplest implementation of a lock key choice we can make
+that covers all the access policy types described above.
+
+Consequences
+************
+
```

Comment:
[clarification] This locking implies that several learners from the same enterprise customer attempting to redeem content using the same policy concurrently won't be possible. In practice, will this lock duration be relatively instantaneous to not block other learners from attempting to enroll via the same policy at the same time as you?

Is the client (MFE) or the API responsible for retrying if a redemption request comes in when the policy is already locked?

Reply:
Changed as follows to answer these questions:
> Locking the entire policy when the policy only applies a per-learner spend cap is fairly restrictive.
  It implies that several learners from the same enterprise customer won't be able to
  concurrently redeem via the same policy.  In practice, this should be a relatively rare occurrence;
  the system throughput is not expected to be particularly high, although the lock duration could
  be on the order of seconds. In this scenario, the client would have to be responsible for any
  retry or communication to the user that a lock could not be acquired.
  We can eventually (or perhaps, immediately) choose to override the lock key choice in the implementing class of 
  that policy type to take the learner identifier into account, which would mitigate the negative effects
  in this scenario.

Reply:
👨‍🍳 💋